src/rust/engine/src/nodes.rs

```diff
@@ -530,25 +528,51 @@ impl Node for SelectTransitive {
       .then(move |dep_product_res| {
         match dep_product_res {
           Ok(dep_product) => {
-            // The product and its dependency list are available: project them.
-            let deps =
-              future::join_all(
-                externs::project_multi(&dep_product, &self.selector.field).iter()
-                  .map(|dep_subject| self.get_dep(&context, &dep_subject))
-                  .collect::<Vec<_>>()
-              );
-            deps
-              .then(move |dep_values_res| {
-                // Finally, store the resulting values.
-                match dep_values_res {
-                  Ok(dep_values) => {
-                    Ok(self.store(&dep_product, dep_values.iter().collect()))
-                  },
-                  Err(failure) =>
-                    Err(was_required(failure)),
-                }
-              })
-              .boxed()
+            let subject_keys = externs::project_multi(&dep_product, &self.selector.field).iter()
+              .map(|subject| externs::key_for(&subject))
+              .collect();
+
+            let init = TransitiveExpansion {
+              todo: subject_keys,
+              outputs: OrderMap::default()
+            };
+
+            loop_fn(init, move |mut expansion| {
+              let round = future::join_all({
+                 expansion.todo.drain()
+                   .map(|subject_key| self.expand_transitive(&context, subject_key))
+                   .collect::<Vec<_>>()
+              });
+
+              round
+                .map(move |finished_items| {
+                  let mut todo_candidates = Vec::new();
+                  for (subject_key, product, more_deps) in finished_items.into_iter() {
+                    expansion.outputs.insert(subject_key, product);
+                    todo_candidates.extend(more_deps);
+                  }
+
+                  // NB mutably borrow expansion.outputs because expansion.to.extens implies mutability.
+                  // enclose with {} to limit the borrowing scope.
+                  {
+                    let outputs = &mut expansion.outputs;
```

Comment:
As discussed offline, this doesn't need to be mutable, although you probably do need to take a reference to it. Also, spelling error above.